📝 Refactor connector breaking changes documentation

aaronsteers opened the pull request:
## What

Reorganizes and enhances connector breaking changes documentation as requested by @aaronsteers in https://app.devin.ai/sessions/644caae698024ce99ebb5a34609b785c.

This PR:
1. Moves breaking changes documentation from `contributing-to-airbyte/` to `connector-development/` directory
2. Renames file to better reflect its focused scope: `connector-breaking-changes.md`
3. Removes redundant general PR contribution process content (~55 lines)
4. Adds detailed guidance on migration guides and metadata requirements based on the internal Breaking Change Release Playbook

## How

**File Reorganization:**
- Deleted: `docs/platform/contributing-to-airbyte/change-cdk-connector.md` (old mixed-purpose file)
- Created: `docs/platform/connector-development/connector-breaking-changes.md` (focused breaking changes doc)
- Moved: `submit-new-connector.md` from `contributing-to-airbyte/` to `connector-development/`

**Content Removed (~55 lines):**
- Entire "Contribution Process" section (issue creation, coding workflow, PR submission, review process)
- This content is covered in the main [Contributing README](https://github.com/airbytehq/airbyte/blob/master/docs/platform/contributing-to-airbyte/README.md)

**Content Added (47 new lines):**
1. **Migration Guide Documentation Requirements** - Explains required WHAT/WHY/WHO/STEPS structure with links to 4 example migration guides
2. **Breaking Change Metadata Requirements** - Details the `message` field (short in-app summary) and `upgradeDeadline` field (7 days minimum for sources, 1 month for destinations)
3. Enhanced "Types of Breaking Changes" with "Full Downstream Rewrite" scenario and Gen-2 approach guidance

**Updated References:**
- PR welcome messages (community + internal) now link to new location with updated title
- Contributing README updated to point to connector development guide
- Added 3 redirects in vercel.json for old paths
- Fixed relative links in new file location (../contributing-to-airbyte/resources/...)

## Review Guide

**🔴 CRITICAL - User-Facing Changes:**
1. `.github/pr-welcome-community.md` (line 8) and `.github/pr-welcome-internal.md` (line 7)
   - These messages appear on EVERY PR for ALL contributors
   - Verify the new title "Managing Breaking Changes in Connectors" is clear and accurate
   - Confirm the URL is correct

2. `docusaurus/vercel.json` (lines 485-497)
   - Verify redirects work for all old paths:
     - `/platform/contributing-to-airbyte/change-cdk-connector` → new location
     - `/platform/contributing-to-airbyte/connector-updates` → new location (from intermediate rename attempt)
     - `/platform/contributing-to-airbyte/submit-new-connector` → new location

**🟡 IMPORTANT - Content Accuracy:**
3. `docs/platform/connector-development/connector-breaking-changes.md` (lines 30-78)
   - **ACCURACY CHECK:** Verify the new guidance matches internal playbook requirements:
     - WHAT/WHY/WHO/STEPS migration guide structure (lines 34-37)
     - Upgrade deadlines: 7 days sources, 1 month destinations (lines 70-72)
     - Exception for immediate upstream breakage (line 76)
   - **LINK VALIDATION:** Verify 4 example migration guide links exist and are good examples (lines 45-48):
     - HubSpot, Google Ads, Stripe, Snowflake migration guides

4. `docs/platform/connector-development/connector-breaking-changes.md` (lines 82-86)
   - **BROKEN LINK RISK:** Verify relative links work from new file location:
     - `../contributing-to-airbyte/resources/pull-requests-handbook.md`
     - `../contributing-to-airbyte/resources/qa-checks.md`

**🟢 MINOR - Organizational Changes:**
5. `docs/platform/contributing-to-airbyte/README.md` (lines 61-65)
   - Verify removed "Contribution Process" content is truly covered in this README or other docs
   - Confirm new connector development guide links are appropriate

## User Impact

**Positive:**
- Contributors making breaking changes have clearer, more actionable guidance with concrete examples
- Migration guide requirements are now explicit (WHAT/WHY/WHO/STEPS structure)
- Upgrade deadline guidance prevents unrealistic timelines
- Better organization: breaking changes doc is now with other connector development docs
- All old URLs redirect automatically (no broken bookmarks)

**Potential Issues:**
- Contributors who bookmarked the old page for general contribution guidance will need to use the main Contributing README instead (redirects handle the URL change, but purpose has shifted)

## Can this PR be safely reverted and rolled back?

- [x] YES 💚

Pure documentation refactoring. Reverting restores old file structure but doesn't break functionality. All changes are in markdown files, redirects, and PR template text.

devin-ai-integration[bot] committed:
📝 Refactor connector changes doc to focus on breaking changes

- Rename 'Changes to Python CDK or Low-Code Connector' to 'Connector Updates'
- Remove basic PR contribution process (covered in main README)
- Keep all breaking changes content as the main focus
- Add 'Related Topics' section with links to all relevant resources
- Update README link text to reflect new focus

This streamlines the documentation to focus specifically on connector
updates and breaking changes, with links to other resources for
general contribution guidance.

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
Rename file to connector-updates.md and add redirects

- Rename change-cdk-connector.md to connector-updates.md
- Add Vercel redirect from old URL to new URL
- Update PR welcome messages (internal and community) to link to new page
- Update README link to point to new filename

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
Revert filename to change-cdk-connector.md for easier review

- Rename back from connector-updates.md to change-cdk-connector.md
- Keep all content changes (removed contribution process, added Related Topics)
- Revert PR welcome message links to use old filename
- Revert vercel.json redirect removal
- Add TODO comment to rename file after PR is merged

This allows GitHub to properly show the file as modified rather than
deleted/created, making the content changes easier to review.

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
Add migration guide and metadata requirements sections

Add detailed guidance for contributors on breaking changes:

1. Migration Guide Requirements section:
   - Explains WHAT/WHY/WHO/STEPS structure from playbook
   - Links to 4 example migration guides (HubSpot, Google Ads, Stripe, Snowflake)
   - Clarifies that guides can be comprehensive (images, SQL, tables)

2. Breaking Change Metadata Requirements section:
   - Message field: Short in-app summary vs detailed migration guide
   - Upgrade deadline field: Minimum timelines (7 days sources, 1 month destinations)
   - Rationale for deadlines and automated notification process

These additions address critical gaps identified when comparing current
docs against the Breaking Change Release Playbook.

Co-Authored-By: AJ Steers <aj@airbyte.io>

aaronsteers committed:
tidy up copy (aj's edits)

aaronsteers committed:
tweak wording (aj)

aaronsteers committed:
move connector stuff under connector development

aaronsteers committed:
tidy up

devin-ai-integration[bot] committed:
Fix redirects and update references after file reorganization

- Add redirects for old paths to new locations in vercel.json:
  * /platform/contributing-to-airbyte/change-cdk-connector → /platform/connector-development/connector-breaking-changes
  * /platform/contributing-to-airbyte/connector-updates → /platform/connector-development/connector-breaking-changes
  * /platform/contributing-to-airbyte/submit-new-connector → /platform/connector-development/submit-new-connector

- Update PR welcome messages (community and internal) to:
  * Use new title 'Managing Breaking Changes in Connectors'
  * Point to new URL /platform/connector-development/connector-breaking-changes
  * Reflect new focus on breaking changes, migration guides, and upgrade deadlines

- Fix broken relative links in connector-breaking-changes.md:
  * Update paths to resources/pull-requests-handbook.md (now ../contributing-to-airbyte/resources/...)
  * Update path to resources/qa-checks.md (now ../contributing-to-airbyte/resources/...)

This addresses file moves from @aaronsteers commits:
- change-cdk-connector.md → connector-breaking-changes.md
- File moved from contributing-to-airbyte/ to connector-development/

Co-Authored-By: AJ Steers <aj@airbyte.io>

devin-ai-integration[bot] committed:
Fix broken links after file reorganization

- Update sidebar-platform.js to remove old paths and add new paths
  * Remove contributing-to-airbyte/change-cdk-connector (moved)
  * Remove contributing-to-airbyte/submit-new-connector (moved)
  * Add connector-development/submit-new-connector
  * Add connector-development/connector-breaking-changes

- Fix broken markdown links in moved files:
  * connector-development/README.md: Update submit-new-connector link
  * contributing-to-airbyte/developing-locally.md: Update submit-new-connector link
  * connector-development/submit-new-connector.md: Fix all asset image paths
  * connector-development/submit-new-connector.md: Fix pull-requests-handbook link

All links now point to correct locations after file moves.
Verified with successful local docs build.

Co-Authored-By: AJ Steers <aj@airbyte.io>

aaronsteers committed:
shorter title in PR welcome message

maxi297 commented on docs/platform/connector-development/connector-breaking-changes.md:
```diff
@@ -0,0 +1,86 @@
+- **Schema Change** - The type of property previously present within a record has changed
+- **Stream or Property Removal** - Data that was previously being synced is no longer going to be synced.
+- **Destination Format / Normalization Change** - The way the destination writes the final data or how normalization cleans that data is changing in a way that requires a full-refresh.
+- **State Changes** - The format of the source’s state has changed, and the full dataset will need to be re-synced
```
We usually use state migration to avoid breaking change on this case

maxi297 commented on docs/platform/connector-development/connector-breaking-changes.md:
```diff
@@ -0,0 +1,86 @@
+A breaking change is any change that requires users to take action before they can continue to sync data. The following are examples of breaking changes:
+
+- **Spec Change** - The configuration required by users of this connector have been changed and syncs will fail until users reconfigure or re-authenticate.  This change is not possible via a Config Migration
+- **Schema Change** - The type of property previously present within a record has changed
```
This is true but I think there were discussions about type compatibility here (if the type change from int to str, this is fine but if it changes from str to int, you better be sure all the str that were there are valid representations of int). For simplicities' sake though, it's probably fine to have this generic rule

aaronsteers replied:
@maxi297 - Agreed. I think we generally think about the distinction between _widened_ types vs narrow or non-compatible types.

- A widening that is still breaking downstream would be from `float` to `str` or `datetime` to `str`.
- A widening that is non-breaking would be from `int` to `bigitn`, or from `int` to `double`.
- A type change from `datetime` to `float` or vice versa is breaking and non-compatible. These types of changes should be avoided if at all possible.

Devin, please hold before applying this change. To fully document might require a dedicated section, and I'm not sure how to state concisely here inline.

maxi297 reviewed (commented):
I used https://whimsical.com/is-change-breaking-DRZqRCsYgGiUHs6weHfYN5 to review this PR.

I also don't remember if changes in PK and cursor are breaking

aaronsteers commented on docs/platform/connector-development/connector-breaking-changes.md:
```diff
@@ -0,0 +1,86 @@
+- **Exception: Immediate upstream breakage:** In the case of immediate upstream breaking changes, such as an already-removed upstream API endpoint, the deadline can be present-day or even in the past - with the rationale that users' connections are _already_ broken without the fix and therefor need the upgrade applied immediately.
+
+- **Automated notifications:** The platform automatically emails users when a breaking change is released and sends reminders as the deadline approaches.
+
```
@maxi297 - I don't know the answer. Perhaps we should log an issue in the enterprise repo? If you give a bit more context, we can ask Devin to log this for us.

maxi297 replied:
I don't. We haven't had a response from expand in the Slack thread

aaronsteers commented on docs/platform/connector-development/submit-new-connector.md:
```diff
@@ -47,7 +47,7 @@ This will enable our team to make sure your contribution does not overlap with e
-3. Follow the [pull request convention](./resources/pull-requests-handbook.md#pull-request-title-convention)
+3. Follow the [pull request convention](../contributing-to-airbyte/resources/pull-requests-handbook.md#pull-request-title-convention)
 4. Wait for a review from a community maintainer or our team. We generally look for the following criteria:
   - Does this PR cover authentication, pagination, and incremental syncs where applicable?
```
aside - 

```suggestion
   - Does this PR cover authentication, pagination, and incremental syncs where applicable?
```

aaronsteers reviewed (commented):
@maxi297 - thanks so much for reviewing this.

(Devin, ignore until we get replies back from @maxi297. Then, you can action those comments if/where they are actionable.)

aaronsteers committed:
Update docs/platform/connector-development/submit-new-connector.md

devin-ai-integration[bot] committed:
Add guidance on migrations and data type compatibility

Based on PR feedback from @maxi297 and @aaronsteers:

1. Add 'Avoiding Breaking Changes with Migrations' section:
   - Config migrations for low-code connectors (with link to CDK schema)
   - State migrations for low-code connectors (with link to CDK schema)
   - Guidance to evaluate migrations before breaking changes

2. Add 'Schema Change Considerations: Data Type Compatibility' section:
   - Widening that is breaking downstream (float→str, datetime→str)
   - Widening that is non-breaking (int→bigint, int→double)
   - Non-compatible type changes (datetime↔float)
   - General guidance to treat type changes as breaking when uncertain

Addresses feedback in https://github.com/airbytehq/airbyte/pull/68098

Co-Authored-By: AJ Steers <aj@airbyte.io>

maxi297 reviewed (approved):
Approving since there is nothing blocking in my comments

aaronsteers commented on docusaurus/vercel.json:
```diff
@@ -480,6 +480,21 @@
+      "source": "/platform/contributing-to-airbyte/connector-updates",
+      "destination": "/platform/connector-development/connector-breaking-changes",
+      "statusCode": 301
+    },
```
aside - 

```suggestion
```

aaronsteers committed:
Update docusaurus/vercel.json
